kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaUtils.java

```diff
@@ -79,6 +83,83 @@ public static void validateSchema(StructType schema, boolean isColumnMappingEnab
     validateSupportedType(schema);
   }
 
+  /**
+   * Performs the following validations on an updated table schema:
+   *
+   * <p>The following checks are performed:
+   *
+   * <ul>
+   *   <li>No duplicate columns are allowed
+   *   <li>Names contain only valid characters
+   *   <li>Data types are supported
+   *   <li>No new non-nullable fields are added
+   *   <li>No type promotions are performed
+   *   <li>Physical column name consistency is preserved in the new schema
+   *   <li>Partition columns are not modified
+   *   <li>Nested IDs for array/map types are preserved in the new schema
+   * </ul>
+   */
+  public static void validateUpdatedSchema(
+      StructType currentSchema,
+      StructType newSchema,
+      List<String> partitionCols,
+      Metadata metadata) {
+    validateSchema(newSchema, true /*columnMappingEnabled*/);
+    ColumnMapping.validateColumnIds(
+        currentSchema, newSchema, TableConfig.ICEBERG_COMPAT_V2_ENABLED.fromMetadata(metadata));
+    validateUpdatedSchemaCompatibility(currentSchema, newSchema, partitionCols);
+  }
+
+  /* Verifies that no non-nullable fields are added, no type promotions are performed and that
+  partition columns are not modified */
+  private static void validateUpdatedSchemaCompatibility(
+      StructType currentSchema, StructType newSchema, List<String> partitionCols) {
+    validatePartitionColumns(newSchema, partitionCols);
+    // Identify added columns based on field IDs
+    Map<Integer, StructField> newSchemaIdToField = idToField(newSchema);
+    Map<Integer, StructField> currentSchemaIdToField = idToField(currentSchema);
+    for (Map.Entry<Integer, StructField> newFieldEntry : newSchemaIdToField.entrySet()) {
+      if (!currentSchemaIdToField.containsKey(newFieldEntry.getKey())) {
+        if (!newFieldEntry.getValue().isNullable()) {
+          throw new KernelException(
+              String.format(
+                  "Cannot add a non-nullable field %s", newFieldEntry.getValue().getName()));
+        }
+      } else {
+        StructField currentField = currentSchemaIdToField.get(newFieldEntry.getKey());
+        StructField newField = newSchemaIdToField.get(newFieldEntry.getKey());
+        if (!newField.getDataType().equivalent(currentField.getDataType())) {
+          throw new KernelException(
+              String.format(
+                  "Cannot change the type of existing field %s from %s to %s",
+                  currentField.getName(), currentField.getDataType(), newField.getDataType()));
```

Comment:
This isn't correct, I need to think more about the right way to do this but this will needleslly prevent legitimate dropping of nested columns because equivalence will fail

Reply:
I'll need to recurse and go field by field explicitly, and account for the fact that struct columns can be dropped. I'll take a look at this with fresh eyes tomorrow

Reply:
I fixed this